dirty_cat/similarity_encoder.py

```diff
@@ -316,11 +394,61 @@ def transform(self, X):
             out = np.empty((len(X), total_length), dtype=self.dtype)
             last = 0
             for j, cats in enumerate(self.categories_):
-                out[:, last:last + len(cats)] = ngram_similarity(X[:, j], cats,
-                                                                 ngram_range=(min_n, max_n),
-                                                                 hashing_dim=self.hashing_dim,
-                                                                 dtype=np.float32)
+                if fast:
+                    encoded_Xj = self._ngram_similarity_fast(X[:, j], j)
+                else:
+                    encoded_Xj = ngram_similarity(
+                        X[:, j], cats, ngram_range=(min_n, max_n),
+                        hashing_dim=self.hashing_dim, dtype=np.float32)
+
+                out[:, last:last + len(cats)] = encoded_Xj
                 last += len(cats)
             return out
         else:
             raise ValueError("Unknown similarity: '%s'" % self.similarity)
+
+    def _ngram_similarity_fast(self, X, col_idx):
+        """
+        Fast computation of ngram similarity, for SimilarityEncoder.
+
+
+        SimilarityEncoder.transform uses the count vectors of the vocabulary in
+        its computations. In ngram_similarity, these count vectors have to be
+        re-computed each time, which can slow down the execution. In this
+        method, the count vectors are recovered from the
+        ``vocabulary_count_matrices`` attribute of the SimilarityEncoder,
+        speeding up the execution.
+        Parameters
+        ----------
+        X: np.array, list
+            observations being transformed.
+        col_idx: int
+            the column index of X in the original feature matrix.
+        """
+        min_n, max_n = self.ngram_range
+        vectorizer = self._vectorizers[col_idx]
+
+        unq_X = np.unique(X)
+        unq_X_ = np.array(list(map(preprocess, unq_X)))
```

Comment:
I wonder whether this line doesn't go faster if you indicate the dtype to np.array. Though it might be negligeable, and thus not worth doing it.

Reply:
In terms of dtype, what do you have in mind? `object`? a fixed-size unicode character length?